src/backends/gpu/stack.hpp

```diff
@@ -22,91 +25,85 @@ namespace gpu {
 template <typename T>
 class stack {
     using value_type = T;
-    using allocator = memory::managed_allocator<value_type>;
-
-    // The number of items of type value_type that can be stored in the stack
-    unsigned capacity_;
-
-    // The number of items that have been stored
-    unsigned size_;
+    template <typename U>
+    using allocator = memory::managed_allocator<U>;
 
-    // Memory containing the value buffer
-    // Stored in managed memory to facilitate host-side access of values
-    // pushed from kernels on the device.
-    value_type* data_;
+    using base_type = stack_base<value_type>;
+    base_type* base_;
 
 public:
+    stack& operator==(const stack& other) = delete;
+    stack(const stack& other) = delete;
 
-    stack(unsigned capacity):
-        capacity_(capacity), size_(0u)
-    {
-        data_ = allocator().allocate(capacity_);
+    stack(stack&& other) {
+        std::swap(base_, other.base_);
     }
 
-    ~stack() {
-        allocator().deallocate(data_, capacity_);
+    stack& operator=(stack&& other) {
+        std::swap(base_, other.base_);
+        return *this;
+    }
+
+    stack(unsigned capacity) {
+        base_ = allocator<base_type>().allocate(1);
+        base_->capacity = capacity;
+        base_->size = 0;
+        base_->data = allocator<value_type>().allocate(capacity);
     }
 
-    // Append a new value to the stack.
-    // The value will only be appended if do_push is true.
-    __device__
-    void push_back(const value_type& value) {
-        // Atomically increment the size_ counter. The atomicAdd returns
-        // the value of size_ before the increment, which is the location
-        // at which this thread can store value.
-        unsigned position = atomicAdd(&size_, 1u);
-
-        // It is possible that size_>capacity_. In this case, only capacity_
-        // entries are stored, and additional values are lost. The size_
-        // will contain the total number of attempts to push,
-        if (position<capacity_) {
-            data_[position] = value;
+    stack(): stack(0) {}
+
```

Comment:
With appropriate checks in the accessors, a default ctor which just sets `base_=0` might be preferable, avoiding any allocations in this case. (Nice fix though for zero-sized `cudaManagedMalloc` case below!)

Reply:
I prefer to have "always valid" state, instead of having two cases in accessors.
The one part of the interface that wouldn't work with that approach is the `base()` method, which has been renamed `storage()` in the latest fixes.

Reply:
Having `base_` always be a valid pointer would also mean that the move constructor needs to make sure that the other `base_` pointer ends up pointing to a valid `stack_base` object. There's a touch of complexity either way, but the always-valid `base_` does involve more allocations.

Reply:
I have gone the route of always valid state. To satisfy the requirements that a moved-from object still be valid, I insist that stacks with capacity 0 and default constructed stacks both point to valid state with `storage.capacity=0`.
This might not be completely optimal, however it gives correct behavior as per the standard, and the type is used in just one location in the back end.